phoenix-core/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java

```diff
@@ -1189,40 +1237,168 @@ private Delete generateDeleteMarkers(List<Cell> row) {
+                        throw new IOException(exceptionMessage);
+                    }
+                    cellCount++;
+                }
+            }
+            if (cellCount != indexRow.rawCells().length) {
+                exceptionMessage = "Index verify failed - Expected to find " + cellCount + " cells but got "
+                        + indexRow.rawCells().length + " cells - " + indexHTable.getName();
+                throw new IOException(exceptionMessage);
+            }
+        }
+
+        private void verifyIndexRows(ArrayList<KeyRange> keys) throws IOException {
+            int expectedRowCount = keys.size();
+            ScanRanges scanRanges = ScanRanges.createPointLookup(keys);
+            Scan indexScan = new Scan();
+            indexScan.setTimeRange(scan.getTimeRange().getMin(), scan.getTimeRange().getMax());
+            scanRanges.initializeScan(indexScan);
+            SkipScanFilter skipScanFilter = scanRanges.getSkipScanFilter();
+            indexScan.setFilter(skipScanFilter);
+            int rowCount = 0;
+            try (ResultScanner resultScanner = indexHTable.getScanner(indexScan)) {
+                for (Result result = resultScanner.next(); (result != null); result = resultScanner.next()) {
+                    Put dataPut = dataPutMap.get(result.getRow());
+                    if (dataPut == null) {
+                        exceptionMessage = "Index verify failed - Missing data row - " + indexHTable.getName();
+                        throw new IOException(exceptionMessage);
+                    }
+                    verifySingleIndexRow(result, dataPut);
+                    rowCount++;
+                }
+            } catch (Throwable t) {
+                ServerUtil.throwIOException(indexHTable.getName().toString(), t);
+            }
+            if (rowCount != expectedRowCount) {
+                exceptionMessage = "Index verify failed - Missing index rows Expected : " + expectedRowCount +
+                        " Actual : " + rowCount + " - " + indexHTable.getName();
+                throw new IOException(exceptionMessage);
+            }
+        }
+
+        private void addVerifyTask(final ArrayList<KeyRange> keys) {
+            tasks.add(new Task<Boolean>() {
+                @Override
+                public Boolean call() throws Exception {
+                    try {
+                        if (Thread.currentThread().isInterrupted()) {
+                            exceptionMessage = "Pool closed, not attempting to verify index rows! " + indexHTable.getName();
+                            throw new IOException(exceptionMessage);
+                        }
+                        verifyIndexRows(keys);
+                    } catch (Exception e) {
+                        throw e;
+                    }
+                    return Boolean.TRUE;
+                }
+            });
+        }
+
+        private void addToBeVerifiedIndexRows() throws IOException {
```

Comment:
nit: do you think moving this method close to the caller makes it more readable?

Reply:
It is called in two places and it is closer to the second one. I will move closer to the first one.